# Evaluation terminale : Machine learning & python -- LIEVRE Mathieu

Cet examen se compose de 3 parties : 
<ol>
  <li>Restitution théorique</li>
  <li>Restitution pratique</li>
  <li>Cas pratique</li>
</ol>

## Partie 1 : Restitution théorique

Pour cette partie, l'objectif est de décrire des notions théoriques. Essayez de rédiger des réponses claires en utilisant vos mots.

#### Question 1.1: Donner une définition du machine learning. Eviter de recopier de wikipédia, essayer d'en donner une description en quelques mots.

Réponse: c'est une IA autodidacte

#### Question 1.2: Donner quelques applications concrètes d'utilisation du machine learning.

Réponse: Le machine learning peut être utilisés dans la médecine (maladie), dans la photo (photos, videos) etc...

#### Question 1.3: Quelles sont les 2 principales familles "d'apprentissage"? Donner une rapide définition de chacune

Réponse: Les deux familles sont:
            - Supervisé: récupéreration des données dés leurs sorties pour entraîner le modèle (label ou une classe cible pour l'algorithme)
            - Non supervisé : L'algorithme crée sois-mémé sa structure

#### Question 1.4: Dans laquelle de ces familles appartient la "classification"? Le "clustering"?

Réponse: ils appartiennent à : 
- Classification: supervisé
- Clustering: non supervisé

#### Question 1.5: Définissez ce qu'est un arbre de décision et décrivez  de manière simple sa construction.

Réponse: Un arbre de décision est une methode de vérification/validité en rapport avec la question : (ex: 20 > 15 ? => oui) Réponse

#### Question 1.6: Dans quelle situation parle-t-on "d'overfitting"?

Réponse: C'est quand le résultat correspond trop à une partie d'un jeu de données spécifiqueRéponse 

## Partie 2 : Etude d'un cas pratique

Pour cette partie, vous devrez vous inspirer des TP réalisés en cours. 

Il vous faut importer les librairies qui vous seront nécessaires répondre à la problèmatique dans la cellule suivante :

In [1]:
# packages
import pandas as pd
#
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier, plot_tree
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.neural_network import MLPClassifier
from sklearn.preprocessing import StandardScaler
import numpy as np
import matplotlib.pyplot as plt

Importez les données comme suit :

In [2]:
data = pd.read_csv('nba_logreg.csv')

Ce jeu de données décrit les caracteristiques de jeu de joueurs de Basketball au sein de la NBA. La variable d'intérêt Y est "TARGET_5yrs". La description de ce dataset se trouve dans le fichier "NBA_variables_doc.png"

__L'objectif du problème est de réussir à prédire si un jeune joueur durera plus ou moins que 5ans au sein de la ligue.__

In [3]:
data

Name  GP   MIN  PTS  FGM  FGA   FG%  3P Made  3PA   3P%  ...  \
0      Brandon Ingram  36  27.4  7.4  2.6  7.6  34.7      0.5  2.1  25.0  ...   
1     Andrew Harrison  35  26.9  7.2  2.0  6.7  29.6      0.7  2.8  23.5  ...   
2      JaKarr Sampson  74  15.3  5.2  2.0  4.7  42.2      0.4  1.7  24.4  ...   
3         Malik Sealy  58  11.6  5.7  2.3  5.5  42.6      0.1  0.5  22.6  ...   
4         Matt Geiger  48  11.5  4.5  1.6  3.0  52.4      0.0  0.1   0.0  ...   
...               ...  ..   ...  ...  ...  ...   ...      ...  ...   ...  ...   
1335      Chris Smith  80  15.8  4.3  1.6  3.6  43.3      0.0  0.2  14.3  ...   
1336      Brent Price  68  12.6  3.9  1.5  4.1  35.8      0.1  0.7  16.7  ...   
1337     Marlon Maxey  43  12.1  5.4  2.2  3.9  55.0      0.0  0.0   0.0  ...   
1338  Litterial Green  52  12.0  4.5  1.7  3.8  43.9      0.0  0.2  10.0  ...   
1339        Jon Barry  47  11.7  4.4  1.6  4.4  36.9      0.4  1.3  33.3  ...   

      FTA   FT%  OREB  DREB  REB  AST  STL  BLK  TOV  TARGET_5Yrs  
0     2.3  69.9   0.7   3.4  4.1  1.9  0.4  0.4  1.3          0.0  
1     3.4  76.5   0.5   2.0  2.4  3.7  1.1  0.5  1.6          0.0  
2     1.3  67.0   0.5   1.7  2.2  1.0  0.5  0.3  1.0          0.0  
3     1.3  68.9   1.0   0.9  1.9  0.8  0.6  0.1  1.0          1.0  
4     1.9  67.4   1.0   1.5  2.5  0.3  0.3  0.4  0.8          1.0  
...   ...   ...   ...   ...  ...  ...  ...  ...  ...          ...  
1335  1.5  79.2   0.4   0.8  1.2  2.5  0.6  0.2  0.8          0.0  
1336  1.0  79.4   0.4   1.1  1.5  2.3  0.8  0.0  1.3          1.0  
1337  1.6  64.3   1.5   2.3  3.8  0.3  0.3  0.4  0.9          0.0  
1338  1.8  62.5   0.2   0.4  0.7  2.2  0.4  0.1  0.8          1.0  
1339  1.0  67.3   0.2   0.7  0.9  1.4  0.7  0.1  0.9          1.0  

[1340 rows x 21 columns]

#### Question 2.1: Donnez au moins une colonne qui doit être enlevé du jeu de données. Pourquoi? 

Réponse: Nous pouvons supprimer par exemple : OREB et DREB => REB, 3P%Réponse 

#### Exercice 2.1: Enlevez les variables du dataset qui vous semble inutile au problème.

In [4]:
# code..
data.drop(columns=['Name'], axis=1, inplace=True)
data

GP   MIN  PTS  FGM  FGA   FG%  3P Made  3PA   3P%  FTM  FTA   FT%  OREB  \
0     36  27.4  7.4  2.6  7.6  34.7      0.5  2.1  25.0  1.6  2.3  69.9   0.7   
1     35  26.9  7.2  2.0  6.7  29.6      0.7  2.8  23.5  2.6  3.4  76.5   0.5   
2     74  15.3  5.2  2.0  4.7  42.2      0.4  1.7  24.4  0.9  1.3  67.0   0.5   
3     58  11.6  5.7  2.3  5.5  42.6      0.1  0.5  22.6  0.9  1.3  68.9   1.0   
4     48  11.5  4.5  1.6  3.0  52.4      0.0  0.1   0.0  1.3  1.9  67.4   1.0   
...   ..   ...  ...  ...  ...   ...      ...  ...   ...  ...  ...   ...   ...   
1335  80  15.8  4.3  1.6  3.6  43.3      0.0  0.2  14.3  1.2  1.5  79.2   0.4   
1336  68  12.6  3.9  1.5  4.1  35.8      0.1  0.7  16.7  0.8  1.0  79.4   0.4   
1337  43  12.1  5.4  2.2  3.9  55.0      0.0  0.0   0.0  1.0  1.6  64.3   1.5   
1338  52  12.0  4.5  1.7  3.8  43.9      0.0  0.2  10.0  1.2  1.8  62.5   0.2   
1339  47  11.7  4.4  1.6  4.4  36.9      0.4  1.3  33.3  0.7  1.0  67.3   0.2   

      DREB  REB  AST  STL  BLK  TOV  TARGET_5Yrs  
0      3.4  4.1  1.9  0.4  0.4  1.3          0.0  
1      2.0  2.4  3.7  1.1  0.5  1.6          0.0  
2      1.7  2.2  1.0  0.5  0.3  1.0          0.0  
3      0.9  1.9  0.8  0.6  0.1  1.0          1.0  
4      1.5  2.5  0.3  0.3  0.4  0.8          1.0  
...    ...  ...  ...  ...  ...  ...          ...  
1335   0.8  1.2  2.5  0.6  0.2  0.8          0.0  
1336   1.1  1.5  2.3  0.8  0.0  1.3          1.0  
1337   2.3  3.8  0.3  0.3  0.4  0.9          0.0  
1338   0.4  0.7  2.2  0.4  0.1  0.8          1.0  
1339   0.7  0.9  1.4  0.7  0.1  0.9          1.0  

[1340 rows x 20 columns]

#### Question 2.2: Ce dataset contient - il des données manquantes ou non exploitable?

Réponse: un certain nombre 

#### Exercice 2.2: Si oui, donnez en la proportion par colonne.

In [5]:
# code..
data.isnull().sum()

GP              0
MIN             0
PTS             0
FGM             0
FGA             0
FG%             0
3P Made         0
3PA             0
3P%            11
FTM             0
FTA             0
FT%             0
OREB            0
DREB            0
REB             0
AST             0
STL             0
BLK             0
TOV             0
TARGET_5Yrs     0
dtype: int64

#### Exercice 2.3: Séparez les données en 2 échantillons train-test

In [6]:
y = data['TARGET_5Yrs']
X = data.drop(columns=['TARGET_5Yrs'], axis=1)

# doing that to try a neural network later on
scaler = StandardScaler()
scaler.fit(X)
X_scaled = scaler.transform(X)

X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size = 0.25, random_state=42)

#### Question 2.3: Pourquoi cette étape est-elle indispensable?

Réponse: Connaitre la pertinance du model

#### Exercice 2.4: Entrainez sur ces données un arbre de décision. Calculer une prédiction.

In [7]:
# code
clf = DecisionTreeClassifier()
clf = clf.fit(X_train, y_train)

estimate = clf.predict(X_test)

print(abs(y_test - estimate).mean())
print((y_test - estimate).std())

ValueError: Input X contains NaN.
DecisionTreeClassifier does not accept missing values encoded as NaN natively. For supervised learning, you might want to consider sklearn.ensemble.HistGradientBoostingClassifier and Regressor which accept missing values encoded as NaNs natively. Alternatively, it is possible to preprocess the data, for instance by using an imputer transformer in a pipeline or drop samples with missing values. See https://scikit-learn.org/stable/modules/impute.html You can find a list of all estimators that handle NaN values at the following page: https://scikit-learn.org/stable/modules/impute.html#estimators-that-handle-nan-values

#### Exercice 2.5: Evaluez la qualité du modèle. 

In [8]:
# code
estimate = clf.predict_proba(X_test)[:,1]

NotFittedError: This DecisionTreeClassifier instance is not fitted yet. Call 'fit' with appropriate arguments before using this estimator.

#### Question 2.4: Quelle métrique utilisez vous? Pourquoi? 

Réponse: simple

#### Question 2.5: Le modèle construit est-il de qualité satisfaisante? Expliquez.

Réponse: La qualité est variable, je ne suis pas assez connaisseur

#### Exercice 2.6: Produisez une visualisation graphique de l'arbre construit

In [9]:
# code
plt.figure(figsize=(17,11))
plot_tree(clf, feature_names = list(X_train.columns), fontsize=10)
plt.show()

AttributeError: 'numpy.ndarray' object has no attribute 'columns'

<Figure size 1700x1100 with 0 Axes>

#### Question 2.6: Donnez un exemple de lecture d'un tel arbre.

Réponse: if GP <= 59.5 alors si FG% <= 39.6 etc...

#### Question 2.7: Comment peut - on améliorer la qualité du modèle? Donnez au moins 3 méthodes:

Réponse: Pour améliorer la qualité du modèle:
            - Optimisation des paramètres

#### Exercice 2.7: Essayer d'appliquer une de ces méthodes

In [10]:
# code
tree = DecisionTreeClassifier(criterion= 'entropy', max_depth=5, min_samples_split=0.1)
tree = tree.fit(X_train, y_train)
tree_predict = tree.predict(X_test)
true = y_test

print(abs(tree_predict - true).mean())
print((tree_predict - true).std())

plt.figure(figsize=(20))
plot_tree(tree, feature_names = list(X_train.columns), fontsize=8)
plt.show()

ValueError: Input X contains NaN.
DecisionTreeClassifier does not accept missing values encoded as NaN natively. For supervised learning, you might want to consider sklearn.ensemble.HistGradientBoostingClassifier and Regressor which accept missing values encoded as NaNs natively. Alternatively, it is possible to preprocess the data, for instance by using an imputer transformer in a pipeline or drop samples with missing values. See https://scikit-learn.org/stable/modules/impute.html You can find a list of all estimators that handle NaN values at the following page: https://scikit-learn.org/stable/modules/impute.html#estimators-that-handle-nan-values

#### Question 2.8: Concluez

Réponse: Je n'arrive pas a bien lancer 